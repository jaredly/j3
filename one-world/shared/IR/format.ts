import ansis from 'ansis';
import { splitGraphemes } from '../../../src/parse/splitGraphemes';
import { Style } from '../nodes';
import { maxLength, white } from './ir-to-text';

const normal = 'abcdefghijklmnopqrstuvwxyz';
const bold = splitGraphemes('𝐚𝐛𝐜𝐝𝐞𝐟𝐠𝐡𝐢𝐣𝐤𝐥𝐦𝐧𝐨𝐩𝐪𝐫𝐬𝐭𝐮𝐯𝐰𝐱𝐲𝐳');
const italic = splitGraphemes('𝑎𝑏𝑐𝑑𝑒𝑓𝑔ℎ𝑖𝑗𝑘𝑙𝑚𝑛𝑜𝑝𝑞𝑟𝑠𝑡𝑢𝑣𝑤𝑥𝑦𝑧');
const bolditalic = splitGraphemes('𝒂𝒃𝒄𝒅𝒆𝒇𝒈𝒉𝒊𝒋𝒌𝒍𝒎𝒏𝒐𝒑𝒒𝒓𝒔𝒕𝒖𝒗𝒘𝒙𝒚𝒛');
const underline = String.fromCharCode(818);

const justify = (text: string) => {
    const lines = text.split('\n');
    const lls = lines.map(ansis.strip);
    const max = maxLength(lls);
    return lines
        .map(
            (l, i) =>
                l + (lls[i].length < max ? white(max - lls[i].length) : ''),
        )
        .join('\n');
};

export const blockFormat = (text: string, style?: Style, enable?: boolean) => {
    if (!style || !enable) return text;
    if (style.background) {
        text = ansis.bgRgb(
            style.background.r,
            style.background.g,
            style.background.b,
        )(justify(text));
        // } else {
        //     text = ansis.bgBlack(text);
    } else if (style.background === false) {
        // Background Base
        // text = ansis.reset(text);
        text = ansis.bgRgb(25, 25, 25)(justify(text));
    }
    if (style.color) {
        text = ansis.rgb(style.color.r, style.color.g, style.color.b)(text);
    }
    return text;
};

export const applyFormats = (text: string, style?: Style, color?: boolean) => {
    if (!style) return text;
    if (style.fontWeight === 'bold') {
        if (style.fontStyle === 'italic') {
            text = convert(text, bolditalic);
        } else {
            text = convert(text, bold);
        }
    } else if (style.fontStyle === 'italic') {
        text = convert(text, italic);
    }
    if (style.textDecoration === 'underline') {
        const emes = splitGraphemes(text);
        text = emes.map((m) => m + underline).join('');
    }
    if (color && style.color) {
        text = ansis.rgb(style.color.r, style.color.g, style.color.b)(text);
    }
    return text;
};

const convert = (text: string, alph: string[]) => {
    let res = '';
    for (let i = 0; i < text.length; i++) {
        const idx = normal.indexOf(text[i]);
        if (idx === -1) res += text[i];
        else res += alph[idx];
    }
    return res;
};
