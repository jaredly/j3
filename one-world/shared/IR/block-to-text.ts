import equal from 'fast-deep-equal';
import { Path, serializePath, Style } from '../nodes';
import { applyFormats, blockFormat, justify } from './format';
import { Block, BlockSource, blockSourceKey, inlineSize } from './ir-to-blocks';
import { addSpaces, joinChunks, white } from './ir-to-text';
import { splitGraphemes } from '../../../src/parse/splitGraphemes';
import ansis from 'ansis';

export type BlockEntry = {
    source: BlockSource;
    shape:
        | {
              type: 'inline';
              start: [number, number];
              end: [number, number];
              hbounds: [number, number]; // min / max X
          }
        | {
              type: 'block';
              start: [number, number];
              width: number;
              height: number;
          };
};

const applyX0 = (text: string, x0: number) => {
    if (x0 === 0) return text;
    const lines = text.split('\n');
    if (lines.length === 1) return text;
    const indent = white(x0);
    return [lines[0], ...lines.slice(1).map((l) => indent + l)].join('\n');
};

export type RGB = { r: number; g: number; b: number };

export type StyleOverrides = Record<
    string,
    | {
          type: 'full';
          color: RGB;
      }
    | {
          type: 'sub';
          color: RGB;
          start: number;
          end: number;
      }
>;

export type DropTarget = {
    path: Path;
    pos: { x: number; y: number };
    side: 'before' | 'after';
    // after: { x: number; y: number };
};

export const blockToText = (
    pos: { x: number; y: number; x0: number },
    block: Block,
    ctx: {
        color: boolean;
        styles: StyleOverrides;
        sourceMaps?: BlockEntry[];
        dropTargets?: DropTarget[];
    },
): string => {
    let nodeStyle = block.style;
    let override = false;
    if (block.node) {
        let key = serializePath(block.node);
        const brace = block.type === 'inline' && block.brace;
        if (brace) key += ':brace';
        const bstyle = ctx.styles[key];
        if (bstyle?.type === 'full') {
            override = true;
            nodeStyle = nodeStyle
                ? { ...nodeStyle, background: bstyle.color }
                : { background: bstyle.color };
        }
        if (ctx.dropTargets && !brace) {
            ctx.dropTargets.push({
                path: block.node,
                side: 'before',
                pos: { x: pos.x, y: pos.y },
            });
            if (block.width > 0) {
                ctx.dropTargets.push({
                    path: block.node,
                    side: 'after',
                    pos: {
                        x: pos.x + block.width,
                        y: pos.y + block.height - 1,
                    },
                });
            }
        }
    }

    switch (block.type) {
        case 'bullet':
            return '▶️ ';
        case 'block': {
            if (block.horizontal !== false) {
                let x = pos.x;
                let h = 0;
                const chunks = block.contents.map((m, i) => {
                    h = Math.max(h, m.height);
                    let y = pos.y;
                    if (block.pullLast && i === block.contents.length - 1) {
                        y = pos.y + h - m.height;
                    }
                    const res = blockToText({ y, x, x0: x }, m, ctx);
                    x +=
                        m.width +
                        (i < block.contents.length - 1
                            ? (block.horizontal as number)
                            : 0);
                    return res;
                });
                if (block.horizontal > 0) {
                    addSpaces(chunks, 'all', white(block.horizontal));
                }
                let text = joinChunks(chunks, block.pullLast);
                if (override) {
                    text = ansis.strip(text);
                }
                return blockFormat(
                    applyX0(text, pos.x - pos.x0),
                    nodeStyle,
                    ctx.color,
                );
            }
            let y = pos.y;
            return blockFormat(
                applyX0(
                    block.contents
                        .map((m) => {
                            const res = blockToText(
                                { ...pos, y, x0: pos.x },
                                m,
                                ctx,
                            );
                            y += m.height;
                            return res;
                        })
                        .join('\n'),
                    pos.x - pos.x0,
                ),
                nodeStyle,
                ctx.color,
            );
        }
        case 'inline': {
            if (block.source && ctx.sourceMaps) {
                ctx.sourceMaps.push({
                    source: block.source,
                    shape:
                        block.height === 1
                            ? {
                                  type: 'block',
                                  start: [pos.x, pos.y],
                                  width: block.width,
                                  height: block.height,
                              }
                            : {
                                  type: 'inline',
                                  start: [pos.x, pos.y],
                                  end: [
                                      pos.x0 + block.last,
                                      pos.y + block.height - 1,
                                  ],
                                  hbounds: [
                                      pos.x0,
                                      Math.max(
                                          pos.x0 + block.width,
                                          pos.x + block.first,
                                      ),
                                  ],
                              },
                });
            }
            const override = block.source
                ? ctx.styles[blockSourceKey(block.source)]
                : undefined;
            if (typeof block.contents === 'string') {
                const chars = splitGraphemes(block.contents);
                if (override?.type === 'sub') {
                    const pre = chars.slice(0, override.start);
                    const mid = chars.slice(override.start, override.end);
                    const post = chars.slice(override.end);
                    if (block.splits) {
                        for (let i = block.splits.length - 1; i >= 0; i--) {
                            const split = block.splits[i];
                            if (split > override.end) {
                                post.splice(split - override.end, 0, '\n');
                            } else if (split > override.start) {
                                mid.splice(split - override.start, 0, '\n');
                            } else {
                                pre.splice(split, 0, '\n');
                            }
                        }
                    }
                    // applySplits
                    return (
                        applyFormats(pre.join(''), nodeStyle, ctx.color) +
                        applyFormats(
                            mid.join(''),
                            { ...nodeStyle, background: override.color },
                            ctx.color,
                        ) +
                        applyFormats(post.join(''), nodeStyle, ctx.color)
                    );
                }

                if (block.splits) {
                    for (let i = block.splits.length - 1; i >= 0; i--) {
                        const split = block.splits[i];
                        chars.splice(split, 0, '\n');
                    }
                }

                if (override?.type === 'full') {
                    return applyFormats(
                        chars.join(''),
                        { ...nodeStyle, background: override.color },
                        ctx.color,
                    );
                }
                return applyFormats(chars.join(''), nodeStyle, ctx.color);
            }
            const style: Style | undefined =
                override?.type === 'full'
                    ? { ...nodeStyle, background: override.color }
                    : nodeStyle;
            let x = pos.x;
            let y = pos.y;
            let x0 = pos.x0;
            return applyFormats(
                block.contents
                    .map((line, i) => {
                        const { width, height, positions, last } = inlineSize(
                            line,
                            x - x0,
                        );
                        const res = line
                            .map((m, i) => {
                                const start = positions[i];
                                const res = blockToText(
                                    {
                                        x: x0 + start.x,
                                        y: y + start.y,
                                        x0: x0,
                                    },
                                    m,
                                    ctx,
                                );
                                if (m.type !== 'inline') {
                                    return justify(res);
                                }
                                return res;
                            })
                            .join('');

                        x = x0;
                        y += height;

                        return res;
                    })
                    .join('\n'),
                style,
                ctx.color,
            );
        }
        case 'table':
            let y = pos.y;
            return blockFormat(
                block.rows
                    .map((row) => {
                        const chunks: string[] = [];
                        let x = pos.x;
                        if (block.sides) {
                            chunks.push('⋮');
                            x++;
                        }
                        let height = 0;
                        for (let i = 0; i < row.length; i++) {
                            if (i > 0) {
                                chunks.push('⋮');
                                x++;
                            }
                            chunks.push(
                                blockToText(
                                    { ...pos, y, x0: x, x },
                                    row[i],
                                    ctx,
                                ),
                            );
                            if (row[i].width < block.colWidths[i]) {
                                chunks.push(
                                    white(block.colWidths[i] - row[i].width),
                                );
                            }
                            x += block.colWidths[i];
                            height = Math.max(height, row[i].height);
                        }
                        if (block.sides) {
                            chunks.push('⋮');
                            x++;
                        }
                        y += height;
                        return joinChunks(chunks);
                    })
                    .join('\n'),
                nodeStyle,
                ctx.color,
            );
    }
};

const highlightStyle = (style?: Style): Style => {
    // console.log('HLL');
    // throw new Error('hk');
    return {
        ...style,
        // color: { r: 255, g: 0, b: 0 },
        // textDecoration: 'underline',
        // fontStyle: 'italic',
        background: { r: 20, g: 50, b: 150 },
    };
};
