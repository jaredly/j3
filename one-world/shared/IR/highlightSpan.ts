import { splitGraphemes } from '../../../src/parse/splitGraphemes';
import { BlockEntry } from './block-to-text';
import { underlineText } from './format';

export const matchesSpan = (
    x: number,
    y: number,
    span: BlockEntry['shape'],
) => {
    if (span.type === 'block') {
        return (
            span.start[0] <= x &&
            span.start[0] + span.width > x &&
            span.start[1] <= y &&
            span.start[1] + span.height > y
        );
    }
    if (y === span.start[1]) {
        return x >= span.start[0] && x <= span.hbounds[1];
    }
    if (y === span.end[1]) {
        return x >= span.hbounds[0] && x <= span.end[0];
    }
    return (
        y > span.start[1] &&
        y < span.end[1] &&
        x >= span.hbounds[0] &&
        x <= span.hbounds[1]
    );
};

export const highlightSpan = (text: string, span: BlockEntry['shape']) => {
    const lines = text.split('\n');
    if (span.type === 'block') {
        return lines
            .map((line, i) => {
                if (
                    i >= span.start[1] &&
                    i <= span.start[1] + span.height - 1
                ) {
                    const chars = splitGraphemes(line);
                    if (span.width === 0 && span.height === 1) {
                        return (
                            chars.slice(0, span.start[0]).join('') +
                            '|' +
                            chars.slice(span.start[0]).join('')
                        );
                    }
                    return (
                        chars.slice(0, span.start[0]).join('') +
                        underlineText(
                            chars
                                .slice(
                                    span.start[0],
                                    span.start[0] + span.width,
                                )
                                .join(''),
                        ) +
                        chars.slice(span.start[0] + span.width).join('')
                    );
                }
                return line;
            })
            .join('\n');
    }
    return lines
        .map((line, i) => {
            if (i >= span.start[1] && i <= span.end[1]) {
                const chars = splitGraphemes(line);
                const start =
                    i === span.start[1] ? span.start[0] : span.hbounds[0];
                const end = i === span.end[1] ? span.end[0] : span.hbounds[1];
                // if (start > end) return line;
                return (
                    chars.slice(0, start).join('') +
                    underlineText(chars.slice(start, end).join('')) +
                    chars.slice(end).join('')
                );
            }
            return line;
        })
        .join('\n');
};
