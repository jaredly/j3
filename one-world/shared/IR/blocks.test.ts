// testingi t up
import { expect, test } from 'bun:test';
import { splitGraphemes } from '../../../src/parse/splitGraphemes';
import { parse } from '../../boot-ex/format';
import { reader } from '../../boot-ex/reader';
import { Control, IR, nodeToIR } from '../../shared/IR/intermediate';
import { LayoutChoices, LayoutCtx, layoutIR } from '../../shared/IR/layout';
import { Nodes, RecNode, Style, toMap } from '../../shared/nodes';
import { BlockEntry, blockToText } from './block-to-text';
import { highlightSpan } from './highlightSpan';
import { Block, irToBlock } from './ir-to-blocks';

const textLayout = (text: string, firstLine: number, style?: Style) => {
    const lines = text.split('\n');
    const height = lines.length;
    const inlineHeight = 1;
    let inlineWidth = firstLine;
    let maxWidth = 0;
    let firstLineWidth = 0;
    lines.forEach((line, i) => {
        inlineWidth = splitGraphemes(line).length;
        if (i === 0) {
            firstLineWidth = inlineWidth + firstLine;
        }
        if (i === 0) inlineWidth += firstLine;
        maxWidth = Math.max(maxWidth, inlineWidth);
    });
    return {
        height,
        inlineHeight,
        inlineWidth,
        maxWidth,
        firstLineWidth,
    };
};

const controlLayout = (control: Control) => {
    let w = 4;
    if (control.type === 'number') {
        w = control.width + 3;
    }
    return {
        height: 1,
        firstLineWidth: w,
        inlineHeight: 1,
        inlineWidth: w,
        maxWidth: w,
    };
};

const processNode = (data: RecNode, maxWidth = 30, leftWidth = 20) => {
    const nodes: Nodes = {};
    const parsed = parse(data);
    const root = toMap(data, nodes);
    const irs: Record<number, IR> = {};
    Object.entries(nodes).forEach(([id, node]) => {
        irs[+id] = nodeToIR(node, parsed.styles, parsed.layouts, {});
    });

    const ctx: LayoutCtx = {
        maxWidth,
        leftWidth,
        irs,
        layouts: {},
        textLayout,
        controlLayout,
    };

    const choices: LayoutChoices = {};
    const result = layoutIR(0, 0, irs[root], choices, ctx);
    ctx.layouts[root] = { choices, result };
    const block = irToBlock(irs[root], irs, choices, {
        space: ',',
        layouts: ctx.layouts,
        annotateNewlines: true,
        top: '',
    });
    const sourceMap: BlockEntry[] = [];
    const txt = blockToText({ x: 0, y: 0, x0: 0 }, block, {
        sourceMaps: sourceMap,
        color: false,
        styles: {},
    });
    return {
        txt,
        result,
        ctx,
        parsed,
        block,
        sourceMap,
    };
};

const process = (text: string, maxWidth = 30, leftWidth = 20) => {
    const data = reader(text, 'a-top');
    if (!data) {
        throw new Error('nothing parsed');
    }
    return processNode(data, maxWidth, leftWidth);
};

const trimTrailingWhite = (txt: string) =>
    txt
        .split('\n')
        .map((t) => t.trimEnd())
        .join('\n');

test('showing spans of simple string', () => {
    expect('\n' + showSpans('"hi"')).toMatchSnapshot();
});

test('showing spans of id', () => {
    expect('\n' + showSpans('hi')).toMatchSnapshot();
});

test('showing spans of string with tag', () => {
    expect('\n' + showSpans('js"hi"')).toMatchSnapshot();
});

test('showing spans of string with inclusion', () => {
    expect('\n' + showSpans('"hi${abc}ho"')).toMatchSnapshot();
});

test('showing spans of wrap string', () => {
    expect(
        '\n' + showSpans('"Hello folks this will wrap, thanks."'),
    ).toMatchSnapshot();
});

test('showing spans of wrap string, short last', () => {
    expect('\n' + showSpans('"Hello folks this will wrap."')).toMatchSnapshot();
});

test('showing spans of wrap string, short last', () => {
    expect(
        '\n' + showSpans('"Hello folks this will ${a} wrap."'),
    ).toMatchSnapshot();
});

test('text with wrap on inclusion', () => {
    expect(
        '\n' + showSpans('"Hello folks ${this} will wrap."'),
    ).toMatchSnapshot();
});

test('highlight over wrap', () => {
    expect('\n' + showSpans('"Hello ${folks} this will."')).toMatchSnapshot();
});

test('showing spans of string with inclusion', () => {
    expect(
        '\n' + showSpans('"hi this is ${a} something ${abc} that will wrap"'),
    ).toMatchSnapshot();
});

// test('showing spans of string with inclusion', () => {
//     expect(
//         '\n' +
//             showSpans(
//                 '(What ok-folks js"abcd${what}here${(what and such js"ok" and stuff)}")',
//             ),
//     ).toMatchSnapshot();
// });

test('simple list', () => {
    expect(
        '\n' + showSpans('(what is)', 100).replaceAll('$', '!'),
    ).toMatchSnapshot();
});

test('list in inclusion', () => {
    expect(
        '\n' + showSpans('"${(what is)}"', 100).replaceAll('$', '!'),
    ).toMatchSnapshot();
});

test('wow ok so bigbug here', () => {
    expect(
        '\n' +
            showSpans(
                'js"sdfgsad${what}asdfjkljsdfk${(loveit and-what"${for}" -what you)}P{} sdfb"',
                100,
            ),
    ).toMatchSnapshot();
    expect(
        '\n' +
            showSpans(
                'js"sdfgsad${what}asdfjkljsdfk${(loveit and-what"${for}" -what you)}P{} sdfb"',
                20,
            ),
    ).toMatchSnapshot();
    expect('\n' + showSpans('"${(lov -sss)}NNN"', 10)).toMatchSnapshot();

    debugger;
    expect(
        '\n' +
            showLayout(
                // STOPSHIP:It should be wrapping before the $(loveit
                // ALSO: The trailing " is way too far out
                '(What js"${10}P n sdfb")',
                10,
            ),
    ).toMatchSnapshot();
    expect(
        '\n' +
            showLayout(
                // STOPSHIP:It should be wrapping before the $(loveit
                // ALSO: The trailing " is way too far out
                '(What js"${10}P nnsdfb")',
                10,
            ),
    ).toMatchSnapshot();

    expect(
        '\n' +
            showSpans(
                // STOPSHIP:It should be wrapping before the $(loveit
                // ALSO: The trailing " is way too far out
                '(What js"${what}asdfaASDJK ${(loveit and-what"${for}" -what you)}P{} sdfb")',
                10,
            ),
    ).toMatchSnapshot();
    expect(
        '\n' +
            showSpans(
                '(What ok-folks hs "" js"sdfasdf${what}asdfaASDJK${(loveit and-what"${for}" -what you)}P{} sdfb")',
                50,
            ),
    ).toMatchSnapshot();
});

test('template string', () => {
    expect('\n' + showSpans('nnnnnnnn"${lol}"')).toMatchSnapshot();
    expect('\n' + showSpans('nnnnnnnnnnnnnnnnn"${lol}"')).toMatchSnapshot();
});

test('small inline', () => {
    let res = [];
    for (let i = 10; i < 20; i++) {
        res.push(showSpans('"One ${two} three"', i));
    }
    expect('\n' + res.join('\n')).toMatchSnapshot();
    // expect('\n' + showSpans('nnnnnnnnnnnnnnnnn"${lol}"')).toMatchSnapshot();
});

const fixDollar = (txt: string) =>
    splitGraphemes(txt)
        .map((t) => (t === '$' ? '!' : t))
        .join('');

export const blockInfo = (block: Block): string => {
    let res = `${block.width}x${block.height}`;
    if (block.type === 'inline') {
        if (typeof block.contents === 'string') {
            res += 'T';
        } else {
            res += `(${block.contents
                .map((line) => line.map(blockInfo).join('|'))
                .join('â†©')})`;
        }
    } else if (block.type === 'block') {
        res += `[${block.contents
            .map(blockInfo)
            .join(block.horizontal === false ? '|' : '-')}]`;
    } else {
        res += '[||]';
    }
    return res;
};

function showLayout(orig: string, width = 20, x = false) {
    let res = Array(width).join('-') + `| ${width}\n`;
    const { txt, block, sourceMap } = process(orig, width);
    res += trimTrailingWhite(fixDollar(txt)) + '\n\n';
    return res;
}

function showSpans(orig: string, width = 20, x = false) {
    let res = Array(width).join('-') + `| ${width}\n`;
    const { txt, block, sourceMap } = process(orig, width);
    res += trimTrailingWhite(fixDollar(txt)) + '\n\n';

    if (x) {
        res += blockInfo(block) + '\n\n';
    }

    sourceMap.forEach((item) => {
        res += `[ # ] ` + JSON.stringify(item.shape) + '\n';
        res +=
            trimTrailingWhite(fixDollar(highlightSpan(txt, item.shape))) + '\n';
    });
    if (x) {
        res += JSON.stringify(block, null, 2);
    }
    return res;
}
