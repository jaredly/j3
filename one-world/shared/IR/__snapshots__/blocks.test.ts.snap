// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`showing spans of simple string 1`] = `
"
"hi"

[ # ] {"type":"block","start":[0,0],"width":0,"height":1}
|"hi"
[ # ] {"type":"block","start":[1,0],"width":2,"height":1}
"h̲i̲"
[ # ] {"type":"block","start":[4,0],"width":0,"height":1}
"hi"|
"
`;

exports[`showing spans of id 1`] = `
"
hi

[ # ] {"type":"block","start":[0,0],"width":2,"height":1}
h̲i̲
"
`;

exports[`showing spans of string with tag 1`] = `
"
js"hi"

[ # ] {"type":"block","start":[0,0],"width":2,"height":1}
j̲s̲"hi"
[ # ] {"type":"block","start":[3,0],"width":2,"height":1}
js"h̲i̲"
[ # ] {"type":"block","start":[6,0],"width":0,"height":1}
js"hi"|
"
`;

exports[`showing spans of string with inclusion 1`] = `
"
"hi＄{abc}ho"

[ # ] {"type":"block","start":[0,0],"width":0,"height":1}
|"hi＄{abc}ho"
[ # ] {"type":"block","start":[1,0],"width":2,"height":1}
"h̲i̲＄{abc}ho"
[ # ] {"type":"block","start":[5,0],"width":3,"height":1}
"hi＄{a̲b̲c̲}ho"
[ # ] {"type":"block","start":[9,0],"width":2,"height":1}
"hi＄{abc}h̲o̲"
[ # ] {"type":"block","start":[12,0],"width":0,"height":1}
"hi＄{abc}ho"|
"
`;

exports[`showing spans of wrap string 1`] = `
"
"Hello folks this  "
 will wrap, thanks.

[ # ] {"type":"block","start":[0,0],"width":0,"height":1}
|"Hello folks this  "
 will wrap, thanks.
[ # ] {"type":"inline","start":[1,0],"end":[19,1],"hbounds":[1,19]}
"H̲e̲l̲l̲o̲ ̲f̲o̲l̲k̲s̲ ̲t̲h̲i̲s̲ ̲ ̲"1:19
 w̲i̲l̲l̲ ̲w̲r̲a̲p̲,̲ ̲t̲h̲a̲n̲k̲s̲.̲1:19
[ # ] {"type":"block","start":[20,0],"width":0,"height":1}
"Hello folks this  "|
 will wrap, thanks.
"
`;

exports[`showing spans of wrap string, short last 1`] = `
"
"Hello folks this "
 will wrap.

[ # ] {"type":"block","start":[0,0],"width":0,"height":1}
|"Hello folks this "
 will wrap.
[ # ] {"type":"inline","start":[1,0],"end":[11,1],"hbounds":[1,18]}
"H̲e̲l̲l̲o̲ ̲f̲o̲l̲k̲s̲ ̲t̲h̲i̲s̲ ̲"1:18
 w̲i̲l̲l̲ ̲w̲r̲a̲p̲.̲1:11
[ # ] {"type":"block","start":[19,0],"width":0,"height":1}
"Hello folks this "|
 will wrap.
"
`;

exports[`showing spans of wrap string, short last 2`] = `
"
"Hello folks this "
 will ＄{a} wrap.

[ # ] {"type":"block","start":[0,0],"width":0,"height":1}
|"Hello folks this "
 will ＄{a} wrap.
[ # ] {"type":"inline","start":[1,0],"end":[6,1],"hbounds":[1,18]}
"H̲e̲l̲l̲o̲ ̲f̲o̲l̲k̲s̲ ̲t̲h̲i̲s̲ ̲"1:18
 w̲i̲l̲l̲ ̲＄{a} wrap.1:6
[ # ] {"type":"block","start":[8,1],"width":1,"height":1}
"Hello folks this "
 will ＄{a̲} wrap.
[ # ] {"type":"block","start":[10,1],"width":6,"height":1}
"Hello folks this "
 will ＄{a} ̲w̲r̲a̲p̲.̲
[ # ] {"type":"block","start":[19,0],"width":0,"height":1}
"Hello folks this "|
 will ＄{a} wrap.
"
`;

exports[`text with wrap on inclusion 1`] = `
"
"Hello folks ＄{this} "
 will wrap.

[ # ] {"type":"block","start":[0,0],"width":0,"height":1}
|"Hello folks ＄{this} "
 will wrap.
[ # ] {"type":"block","start":[1,0],"width":12,"height":1}
"H̲e̲l̲l̲o̲ ̲f̲o̲l̲k̲s̲ ̲＄{this} "
 will wrap.
[ # ] {"type":"block","start":[15,0],"width":4,"height":1}
"Hello folks ＄{t̲h̲i̲s̲} "
 will wrap.
[ # ] {"type":"inline","start":[20,0],"end":[11,1],"hbounds":[1,11]}
"Hello folks ＄{this}s ＄{this} "20:11
 w̲i̲l̲l̲ ̲w̲r̲a̲p̲.̲1:11
[ # ] {"type":"block","start":[22,0],"width":0,"height":1}
"Hello folks ＄{this} "|
 will wrap.
"
`;
